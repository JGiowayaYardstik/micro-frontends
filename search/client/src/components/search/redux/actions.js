const createActionName = name => `search/search/${name}`;
export const UPDATE_SEARCH_DATA = createActionName('UPDATE_SEARCH_DATA');
export const UPDATE_COUNT_DATA = createActionName('UPDATE_COUNT_DATA');
export const UPDATE_SUGGESTIONS = createActionName('UPDATE_SUGGESTIONS');
export const SEARCH_ERROR = createActionName('SEARCH_ERROR');
export const SEARCH_COUNT_ERROR = createActionName('SEARCH_COUNT_ERROR');
export const SET_LOADING_STATE = createActionName('SET_LOADING_STATE');
export const SET_DISPLAY_SUGGESTIONS = createActionName('SET_DISPLAY_SUGGESTIONS');
export const SET_ACTIVE_SEARCH_PROVIDER = createActionName('SET_ACTIVE_SEARCH_PROVIDER');
export const RESET_SEARCH_STATE = createActionName('RESET_SEARCH_STATE');

const updateSearchData = (ID, activeSearchProvider, queryData, data) => ({
	type: UPDATE_SEARCH_DATA,
	ID,
	activeSearchProvider,
	queryData,
	data
});

const updateCountData = (ID, data) => ({
	type: UPDATE_COUNT_DATA,
	ID,
	data
});

const updateSuggestions = (ID, data) => ({
	type: UPDATE_SUGGESTIONS,
	ID,
	data
});

const searchError = (ID, activeSearchProvider, message) => ({
	type: SEARCH_ERROR,
	ID,
	activeSearchProvider,
	message
});

const searchCountError = (ID, message) => ({
	type: SEARCH_COUNT_ERROR,
	ID,
	message
});

const setLoadingState = isLoading => ({
	type: SET_LOADING_STATE,
	isLoading
});

const setDisplaySuggestions = displaySuggestions => ({
	type: SET_DISPLAY_SUGGESTIONS,
	displaySuggestions
});

const setActiveSearchProvider = activeSearchProvider => ({
	type: SET_ACTIVE_SEARCH_PROVIDER,
	activeSearchProvider
});

const resetSearchState = () => ({
	type: RESET_SEARCH_STATE
});

export {
	updateSearchData,
	updateCountData,
	updateSuggestions,
	searchError,
	searchCountError,
	setLoadingState,
	setDisplaySuggestions,
	setActiveSearchProvider,
	resetSearchState
};
