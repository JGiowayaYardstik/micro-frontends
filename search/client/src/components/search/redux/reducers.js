import { createStructuredSelector } from 'reselect';
import { getSearchProviders } from 'search-api';

import {
	UPDATE_SEARCH_DATA,
	UPDATE_COUNT_DATA,
	SEARCH_ERROR,
	SEARCH_COUNT_ERROR,
	SET_LOADING_STATE,
	UPDATE_SUGGESTIONS,
	SET_DISPLAY_SUGGESTIONS,
	SET_ACTIVE_SEARCH_PROVIDER,
	RESET_SEARCH_STATE
} from './actions';
import isEmpty from '../../../util/isEmpty';

export const reducerNameSearch = 'search/search';

const hasNoValidResult = current => {
	if (isEmpty(current.data)) {
		return true;
	}

	return !current.data.count || current.data.count === 0;
};

// export only for testing
export const getValidActiveProvider = (current, searchProviders, searchState) => {
	const isCurrentProviderActive = current && current.activeProvider.ID === current.ID;

	if (isCurrentProviderActive && hasNoValidResult(current)) {
		const newSearchProviders = searchProviders.filter(provider => provider.ID !== current.ID);
		const newActive = { ...newSearchProviders[0] };
		const newActiveData = searchState[newActive.ID];

		if (!isEmpty(newSearchProviders)) {
			return getValidActiveProvider(
				{
					ID: newActive.ID,
					data: newActiveData,
					activeProvider: newActive
				},
				newSearchProviders,
				searchState
			);
		}
		// if there is no provider with any search results,
		// we return null and show search suggestions
		return null;
	}
	return current ? current.activeProvider : null;
};

const updateSearchData = (action, state) => {
	const { ID, activeSearchProvider, queryData, data } = action;

	const newSearchState = {
		...state.searchState
	};

	newSearchState[ID] = data;

	const newQueryState = {
		...state.queryState
	};

	newQueryState[ID] = {
		...queryData,
		searchType: ID.toLowerCase()
	};

	const newErrorState = {
		...state.errorState
	};

	newErrorState[ID] = { error: false, message: '' };

	return {
		...state,
		activeSearchProvider: getValidActiveProvider(
			{ ID, data, activeProvider: { ...activeSearchProvider } },
			getSearchProviders(),
			newSearchState
		),
		searchState: newSearchState,
		errorState: newErrorState,
		queryState: newQueryState
	};
};

const updateSuggestions = (action, state) => {
	const { ID, data } = action;

	const newSearchState = {
		...state.searchState
	};

	newSearchState[ID] = data;

	return {
		...state,
		activeSearchProvider: null,
		searchState: newSearchState
	};
};

const updateCountData = (action, state) => {
	const { ID, data } = action;

	const newSearchState = {
		...state.searchState
	};

	newSearchState[ID] = {
		...state.searchState[ID],
		count: data.count
	};

	const newErrorState = {
		...state.errorState
	};

	newErrorState[ID] = { error: false, message: '' };

	return {
		...state,
		searchState: newSearchState,
		errorState: newErrorState
	};
};

const searchError = (action, state) => {
	const { ID, activeSearchProvider } = action;

	const newSearchState = {
		...state.searchState
	};

	newSearchState[ID] = {};

	const newErrorState = {
		...state.errorState
	};

	newErrorState[ID] = { error: true, message: action.message };

	return {
		...state,
		activeSearchProvider: getValidActiveProvider(
			{ ID, data: {}, activeProvider: { ...activeSearchProvider } },
			getSearchProviders(),
			newSearchState
		),
		searchState: newSearchState,
		errorState: newErrorState,
		loadingState: false
	};
};

const searchCountError = (action, state) => {
	const { ID } = action;

	const newSearchState = {
		...state.searchState
	};

	newSearchState[ID] = {};

	const newErrorState = {
		...state.errorState
	};

	newErrorState[ID] = { error: true, message: action.message };

	return {
		...state,
		searchState: newSearchState,
		errorState: newErrorState
	};
};

const initialState = {
	searchState: {},
	errorState: {},
	queryState: {},
	activeSearchProvider: {},
	loadingState: false,
	displaySuggestions: false
};

const search = (state = initialState, action) => {
	switch (action.type) {
		case UPDATE_SEARCH_DATA:
			return updateSearchData(action, state);

		case UPDATE_COUNT_DATA:
			return updateCountData(action, state);

		case UPDATE_SUGGESTIONS:
			return updateSuggestions(action, state);

		case SEARCH_ERROR:
			return searchError(action, state);

		case SEARCH_COUNT_ERROR:
			return searchCountError(action, state);

		case SET_LOADING_STATE:
			return {
				...state,
				loadingState: action.isLoading
			};

		case SET_DISPLAY_SUGGESTIONS:
			return {
				...state,
				displaySuggestions: action.displaySuggestions
			};

		case SET_ACTIVE_SEARCH_PROVIDER:
			return {
				...state,
				activeSearchProvider: action.activeSearchProvider
			};

		case RESET_SEARCH_STATE:
			return initialState;

		default:
			return state;
	}
};

const selectSearchState = state => state[reducerNameSearch] && state[reducerNameSearch].searchState;
const selectErrorState = state => state[reducerNameSearch] && state[reducerNameSearch].errorState;
const selectQueryState = state => state[reducerNameSearch] && state[reducerNameSearch].queryState;
const selectLoadingState = state =>
	state[reducerNameSearch] && state[reducerNameSearch].loadingState;
const selectDisplaySuggestions = state =>
	state[reducerNameSearch] && state[reducerNameSearch].displaySuggestions;
const selectActiveSearchProvider = state =>
	state[reducerNameSearch] && state[reducerNameSearch].activeSearchProvider;

export const structuredSelector = createStructuredSelector({
	searchState: selectSearchState,
	errorState: selectErrorState,
	queryState: selectQueryState,
	loadingState: selectLoadingState,
	displaySuggestions: selectDisplaySuggestions,
	activeSearchProvider: selectActiveSearchProvider
});

global.reducerRegistry.register(reducerNameSearch, search);

export default search;
