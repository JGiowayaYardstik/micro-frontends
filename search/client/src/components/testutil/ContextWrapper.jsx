import React from 'react';
import PropTypes from 'prop-types';
import { StyleProvider } from '@dm/style-provider';
import { DeviceProvider } from '@dm/device-provider';
import { I18nextProvider } from 'react-i18next';
import { Provider } from 'react-redux';
import { theme as dm } from '@dm/ui-theme';

export default function ContextWrapper({ children, locale }) {
	return (
		<StyleProvider>
			<DeviceProvider>
				<I18nextProvider i18n={global.i18next} initialLanguage={locale}>
					{children}
				</I18nextProvider>
			</DeviceProvider>
		</StyleProvider>
	);
}

ContextWrapper.propTypes = {
	children: PropTypes.node,
	locale: PropTypes.string
};

export function ContextWrapperWithoutResponsive({ children, locale }) {
	return (
		<StyleProvider>
			<I18nextProvider i18n={global.i18next} initialLanguage={locale}>
				{children}
			</I18nextProvider>
		</StyleProvider>
	);
}

ContextWrapperWithoutResponsive.propTypes = {
	children: PropTypes.node,
	locale: PropTypes.string
};

export function ContextWrapperWithStore({ children, locale, theme = dm, store = global.store }) {
	return (
		<Provider store={store}>
			<StyleProvider theme={theme}>
				<I18nextProvider i18n={global.i18next} initialLanguage={locale}>
					{children}
				</I18nextProvider>
			</StyleProvider>
		</Provider>
	);
}

ContextWrapperWithStore.propTypes = {
	children: PropTypes.node,
	locale: PropTypes.string,
	store: PropTypes.object,
	theme: PropTypes.object
};
