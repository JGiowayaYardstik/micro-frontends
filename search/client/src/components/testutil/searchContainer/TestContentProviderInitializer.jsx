import React from 'react';
import { registerContentProvider } from 'search-api';

const initializeTestContentProvider = () => {
  registerContentProvider({
    getContentComponent: contentId => <div>content component {contentId}</div>,
  });
};

export default initializeTestContentProvider;
