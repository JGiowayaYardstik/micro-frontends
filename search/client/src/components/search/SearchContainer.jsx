/* eslint import/prefer-default-export: 0 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Loading from '@dm/loading';
import { compose } from 'recompose';
import { connect as connectFela } from '@dm/style-provider';
import { withGlobalConfig } from '@dm/global-config-provider';
import { getSearchProviders } from 'search-api';
import SearchResultHeader from './SearchResultHeader';
import SearchResultBody from './SearchResultBody';
import SearchNavigation from './SearchNavigation';
import {
  doLoadData,
  doLoadCount,
  doLoadSuggestions,
  setActiveSearchProvider,
} from './redux/actions';
import { structuredSelector as mapStateToProps } from './redux/reducers';
import isEmpty from '../../util/isEmpty';
import SearchInput from './SearchInput';

export class UnconnectedSearchContainer extends Component {
  componentDidMount() {
    this.props.setActiveSearchProvider(getSearchProviders()[0]);
    window.addEventListener('popstate', this.handleHistoryPopState);

    this.unsubscribeListeners = () => {
      window.removeEventListener('popstate', this.handleHistoryPopState);
    };
  }

  componentWillUnmount() {
    this.unsubscribeListeners();
  }

  onTabClick = searchProvider => {
    if (searchProvider.handlePushHistory) {
      searchProvider.handlePushHistory(this.getQueryData(searchProvider));
    }
    this.handleSearch(searchProvider.ID, this.getQueryData(searchProvider));
  };

  getActiveSearchProvider = searchType => {
    const searchProviders = getSearchProviders();

    const activeSearchProviderFromSearchType = searchProviders.filter(
      searchProvider => searchType && searchProvider.ID === searchType.toUpperCase()
    )[0];

    return activeSearchProviderFromSearchType || searchProviders[0];
  };

  getQueryData = searchProvider => {
    const { queryState } = this.props;
    return !isEmpty(queryState) && searchProvider && !isEmpty(queryState[searchProvider.ID])
      ? queryState[searchProvider.ID]
      : this.props.queryData;
  };

  fetchData = queryData => {
    const searchType = queryData ? queryData.searchType : '';
    const activeSearchProvider = this.getActiveSearchProvider(searchType);
    this.handleSearch(activeSearchProvider.ID, queryData);
  };

  handleUpdateSuggestions = () => {
    if (!this.props.displaySuggestions) {
      const searchProviders = getSearchProviders();

      searchProviders.forEach(searchProvider =>
        this.props.loadSuggestions({
          searchProviderId: searchProvider.ID,
          queryData: this.props.queryData,
        })
      );
    }
  };

  handleSearch = (activeSearchProviderID, queryData) => {
    const searchProviders = getSearchProviders();

    searchProviders.forEach(provider => {
      this.props.loadCount({
        searchProviderId: provider.ID,
        providers: searchProviders,
        queryData,
      });
    });

    this.props.loadData({
      searchProviderId: activeSearchProviderID,
      queryData,
      providers: searchProviders,
    });
  };

  handleHistoryPopState = event => {
    const newState = event.state;

    if (newState) {
      const searchType = newState.searchType;
      if (searchType && searchType.toUpperCase() !== this.props.activeSearchProvider.ID) {
        const activeSearchProvider = this.getActiveSearchProvider(searchType);
        this.props.setActiveSearchProvider(activeSearchProvider);
      }
    } else {
      const searchProviders = getSearchProviders();
      this.props.setActiveSearchProvider(searchProviders[0]);
    }
  };

  render() {
    const {
      searchFallback,
      activeSearchProvider,
      searchState,
      countState,
      errorState,
      loadingState,
      styles,
    } = this.props;

    return (
      <div>
        <SearchInput
          onSubmit={query => this.handleSearch(activeSearchProvider.ID, query)}
          value={
            this.getQueryData(activeSearchProvider) && this.getQueryData(activeSearchProvider).query
          }
        />
        {activeSearchProvider && this.getQueryData(activeSearchProvider) && (
          <div data-dmid="search-container" className={styles.searchContainer}>
            {activeSearchProvider.getNavComponent && (
              <SearchNavigation
                activeSearchProvider={activeSearchProvider}
                queryData={this.getQueryData(activeSearchProvider)}
                searchState={searchState}
                fetchData={this.fetchData}
              />
            )}
            <div data-dmid="search-content-container" className={styles.searchContentContainer}>
              <SearchResultHeader
                query={this.getQueryData(activeSearchProvider).query}
                activeSearchProvider={activeSearchProvider}
                countState={countState}
                onTabClick={searchProvider => this.onTabClick(searchProvider)}
              />

              <SearchResultBody
                activeSearchProvider={activeSearchProvider}
                fetchData={this.fetchData}
                errorState={errorState}
                queryData={this.getQueryData(activeSearchProvider)}
                searchState={searchState}
                searchFallback={searchFallback}
                handleUpdateSuggestions={this.handleUpdateSuggestions}
              />
            </div>
            {loadingState && <Loading />}
          </div>
        )}
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  loadData: (...args) => doLoadData(...args)(dispatch),
  loadCount: (...args) => doLoadCount(...args)(dispatch),
  loadSuggestions: (...args) => doLoadSuggestions(...args)(dispatch),
  setActiveSearchProvider: (...args) => dispatch(setActiveSearchProvider(...args)),
});

const searchContainer = ({ theme }) => ({
  position: 'relative',
  display: 'flex',
  maxWidth: theme.dimension.contentSize.l.px,
  margin: `${theme.dimension.spacing.xl.rem} auto 0`,
  padding: `0 ${theme.dimension.spacing.xs.rem}`,
  column2to6: {
    margin: `${theme.dimension.spacing.xs.rem} auto 0`,
    padding: '0',
  },
});

const searchContentContainer = () => ({
  flex: '1 0 0px',
});

UnconnectedSearchContainer.propTypes = {
  // fela props
  styles: PropTypes.object,
  // own props
  searchFallback: PropTypes.object,
  queryData: PropTypes.object.isRequired,
  // props from store
  searchState: PropTypes.object,
  countState: PropTypes.object,
  errorState: PropTypes.object,
  queryState: PropTypes.object,
  loadingState: PropTypes.bool,
  displaySuggestions: PropTypes.bool,
  activeSearchProvider: PropTypes.object,
  loadData: PropTypes.func.isRequired,
  loadCount: PropTypes.func.isRequired,
  loadSuggestions: PropTypes.func.isRequired,
  setActiveSearchProvider: PropTypes.func.isRequired,
  // ConfigProvider
  getConfig: PropTypes.func, // eslint-disable-line react/no-unused-prop-types
};

export default compose(
  withGlobalConfig,
  connect(
    mapStateToProps,
    mapDispatchToProps
  ),
  connectFela({ searchContainer, searchContentContainer })
)(UnconnectedSearchContainer);
