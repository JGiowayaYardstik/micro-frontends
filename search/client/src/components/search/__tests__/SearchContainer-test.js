import React from 'react';
import { Provider } from 'react-redux';
import { createSnapshot, createMockStore } from '@dm/jest-utils';
import { mount, shallow } from 'enzyme';
import axios from 'axios';
import ContextWrapper, { ContextWrapperWithStore } from '../../testutil/ContextWrapper';
import * as actions from '../redux/actions';
import { reducerNameSearch } from '../redux/reducers';

import {
	initialize as initializeTestSearch,
	shutdown as shutDownTestProvider,
	testProvider
} from '../../testutil/searchContainer/TestSearchInitializer';
import initializeTestContentProvider from '../../testutil/searchContainer/TestContentProviderInitializer';
import { searchResults as results } from '../../testutil/searchContainer/actions';

import SearchContainer, { UnconnectedSearchContainer } from '../SearchContainer';

const getProps = query => ({
	locale: 'de',
	site: 'de_mcr',
	targetSystem: '',
	servicesGatewayUrl: 'https://services-stage.dm-drogeriemarkt.com',
	query: query || 'bio',
	searchType: 'product',
	parametersProductSearch: {
		categoryId: '',
		initialQuery: '',
		productQuery: '',
		hiddenFacets: '',
		currentPage: 1,
		pageSize: 24,
		sort: 'relevance'
	}
});

const testProviderProduct = testProvider('PRODUCT', 1);
const testProviderContent = testProvider('CONTENT', 2);

const store = createMockStore({
	[reducerNameSearch]: {
		searchState: {
			CONTENT: { results },
			PRODUCT: { results }
		},
		errorState: {},
		activeSearchProvider: { testProviderProduct },
		searchType: 'product',
		queryData: { ...getProps }
	}
});

describe('SearchContainer', () => {
	beforeEach(() => {
		jest.spyOn(window.history, 'pushState').mockImplementation(() => {
			/* do nothing */
		});
		jest.spyOn(window, 'addEventListener').mockImplementation(() => {
			/* do nothing */
		});
		jest.spyOn(window, 'removeEventListener').mockImplementation(() => {
			/* do nothing */
		});
	});

	afterEach(() => {
		store.clearActions();
		jest.clearAllMocks();
	});

	it('renders and matches snapshot when results are avaibable', () => {
		initializeTestSearch(testProviderProduct);
		initializeTestSearch(testProviderContent);

		const props = setupDefaultSearchContainerProperties();

		const snapshot = createSnapshot(
			<ContextWrapper>
				<UnconnectedSearchContainer {...props} />
			</ContextWrapper>
		);

		expect(snapshot).toMatchSnapshot();
	});

	it('renders and matches snapshot when no results are found (search fallback)', () => {
		initializeTestSearch(testProviderProduct);
		shutDownTestProvider('CONTENT');
		initializeTestContentProvider();

		const snapshot = createSnapshot(
			<ContextWrapper>
				<UnconnectedSearchContainer
					queryData={getProps('2ß9358ß239648')}
					searchType="product"
					searchFallback={{
						textId1: 535460,
						textId2: 535470
					}}
					setLoadingState={() => {}}
					setDisplaySuggestions={() => {}}
					setActiveSearchProvider={() => {}}
					updateSearchData={() => {}}
					updateCountData={() => {}}
					updateSuggestions={() => {}}
					searchError={() => {}}
					searchCountError={() => {}}
					searchState={{
						PRODUCT: {
							count: 0,
							results: [],
							suggestions: []
						}
					}}
					activeSearchProvider={null}
					styles={{}}
				/>
			</ContextWrapper>
		);
		expect(snapshot).toMatchSnapshot();
	});

	it('mounts and checks if container elements are available', () => {
		store.clearActions();
		initializeTestSearch(testProviderProduct);
		initializeTestSearch(testProviderContent);
		const props = setupDefaultSearchContainerProperties();

		const component = mount(
			<ContextWrapperWithStore store={store}>
				<ContextWrapper>
					<UnconnectedSearchContainer {...props} />
				</ContextWrapper>
			</ContextWrapperWithStore>
		);

		expect(component.find('[data-dmid="search-container"]')).toHaveLength(1);
		expect(component.find('[data-dmid="search-header"]')).toHaveLength(1);
		expect(component.find('[data-dmid="search-nav"]')).toHaveLength(1);
		expect(component.find('[data-dmid="search-tab-PRODUCT"]')).toHaveLength(1);
		expect(component.find('[data-dmid="search-tab-CONTENT"]')).toHaveLength(1);
		expect(component.find('[data-dmid="search-body"]')).toHaveLength(1);
	});

	it('mounts and updates search data (active search)', () => {
		initializeTestSearch(testProviderProduct);
		initializeTestSearch(testProviderContent);

		// expected update Search Data (active search provider)
		const expectedUpdateSearchData = {
			type: actions.UPDATE_SEARCH_DATA,
			ID: 'PRODUCT',
			activeSearchProvider: testProviderProduct,
			data: {
				searchResults: results
			},
			queryData: getProps()
		};

		axios.get = jest.fn().mockImplementation(
			() =>
				new Promise((resolve, reject) => {
					reject(store.dispatch(expectedUpdateSearchData));
				})
		);

		function updateSearchData() {
			return new Promise(resolve => setImmediate(resolve(results)));
		}

		mount(
			<ContextWrapperWithStore store={store}>
				<SearchContainer
					queryData={getProps()}
					searchType="product"
					searchFallback={{
						textId1: 535460,
						textId2: 535470
					}}
				/>
			</ContextWrapperWithStore>
		);

		return updateSearchData().then(() => {
			expect(
				store.getActions().filter(action => action.type === actions.UPDATE_SEARCH_DATA)
			).toHaveLength(1);
			expect(
				store.getActions().filter(action => action.type === actions.UPDATE_SEARCH_DATA)[0]
			).toEqual(expectedUpdateSearchData);
		});
	});

	it('mounts and updates count data (inactive search tab)', () => {
		initializeTestSearch(testProviderProduct);
		initializeTestSearch(testProviderContent);

		// expected update Count Data (inactive search provider)
		const expectedUpdateCountData = {
			type: actions.UPDATE_COUNT_DATA,
			ID: 'CONTENT',
			data: {
				searchResults: results
			}
		};

		axios.get = jest.fn().mockImplementation(
			() =>
				new Promise((resolve, reject) => {
					reject(store.dispatch(expectedUpdateCountData));
				})
		);

		function updateSearchData() {
			return new Promise(resolve => setImmediate(resolve(results)));
		}

		mount(
			<ContextWrapperWithStore store={store}>
				<SearchContainer
					queryData={getProps()}
					searchType="product"
					searchFallback={{
						textId1: 535460,
						textId2: 535470
					}}
				/>
			</ContextWrapperWithStore>
		);

		return updateSearchData().then(() => {
			expect(
				store.getActions().filter(action => action.type === actions.UPDATE_COUNT_DATA)
			).toHaveLength(1);
			expect(
				store.getActions().filter(action => action.type === actions.UPDATE_COUNT_DATA)[0]
			).toEqual(expectedUpdateCountData);
		});
	});

	it('registers history popstate handler on component mount', () => {
		const props = setupDefaultSearchContainerProperties();
		expect(window.addEventListener).not.toHaveBeenCalled();
		const wrapper = shallow(<UnconnectedSearchContainer {...props} />);
		expect(window.addEventListener).toHaveBeenCalledTimes(1);
		expect(window.addEventListener).toHaveBeenCalledWith(
			'popstate',
			wrapper.instance().handleHistoryPopState
		);
	});

	it('unsubscribes history popstate handler on component unmount', () => {
		const props = setupDefaultSearchContainerProperties();
		const wrapper = shallow(<UnconnectedSearchContainer {...props} />);
		expect(window.removeEventListener).not.toHaveBeenCalled();
		wrapper.instance().componentWillUnmount();
		expect(window.removeEventListener).toHaveBeenCalledTimes(1);
		expect(window.removeEventListener).toHaveBeenCalledWith(
			'popstate',
			wrapper.instance().handleHistoryPopState
		);
	});

	it('sets new active search provider when searchType is toggled', () => {
		const expectedSearchProvider = testProviderContent;

		const props = setupDefaultSearchContainerProperties();
		const wrapper = shallow(
			<UnconnectedSearchContainer {...props} activeSearchProvider={testProviderProduct} />
		);

		const popstateEvent = {
			state: {
				// type is lowercase whereas provider.ID is uppercase!
				type: expectedSearchProvider.ID.toLowerCase()
			}
		};
		wrapper.instance().handleHistoryPopState(popstateEvent);

		expect(props.setActiveSearchProvider).toHaveBeenCalledWith(expectedSearchProvider);
	});

	it('does not set active search provider when searchType is the same', () => {
		const popstateEvent = {
			state: {
				// type is lowercase whereas provider.ID is uppercase!
				type: testProviderProduct.ID.toLowerCase()
			}
		};
		ensureHistoryPopStateDoesNotChangeActiveSearchProvider(popstateEvent);
	});

	it('does not set active search provider when searchType undefined', () => {
		const popstateEvent = {
			state: {
				type: undefined
			}
		};
		ensureHistoryPopStateDoesNotChangeActiveSearchProvider(popstateEvent);
	});

	it('set default active search provider if history state is undefined', () => {
		const mockGetSearchProviders = jest.fn();
		jest.mock('search-api', () => ({ getSearchProviders: mockGetSearchProviders }));
		const props = setupDefaultSearchContainerProperties();
		shallow(<UnconnectedSearchContainer {...props} />);
		expect(props.setActiveSearchProvider).toHaveBeenCalledWith(testProviderProduct);
	});

	function ensureHistoryPopStateDoesNotChangeActiveSearchProvider(popstateEvent) {
		const props = setupDefaultSearchContainerProperties();
		// disableLifecycleMethods is needed to isolate handleHistoryPopState()
		const wrapper = shallow(
			<UnconnectedSearchContainer {...props} activeSearchProvider={testProviderProduct} />,
			{ disableLifecycleMethods: true }
		);
		wrapper.instance().handleHistoryPopState(popstateEvent);
		expect(props.setActiveSearchProvider).not.toHaveBeenCalled();
	}

	function setupDefaultSearchContainerProperties() {
		return {
			queryData: getProps(),
			queryState: {
				PRODUCT: getProps(),
				CONTENT: getProps()
			},
			searchType: 'product',
			searchFallback: {
				textId1: 535460,
				textId2: 535470
			},
			searchState: {
				CONTENT: { results },
				PRODUCT: { results }
			},
			activeSearchProvider: testProviderProduct,
			styles: {},
			setLoadingState: jest.fn(),
			setDisplaySuggestions: jest.fn(),
			setActiveSearchProvider: jest.fn(),
			updateSearchData: jest.fn(),
			updateCountData: jest.fn(),
			updateSuggestions: jest.fn(),
			searchError: jest.fn(),
			searchCountError: jest.fn()
		};
	}
});
