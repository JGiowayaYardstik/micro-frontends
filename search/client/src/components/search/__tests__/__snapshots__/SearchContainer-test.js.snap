// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchContainer renders and matches snapshot when no results are found (search fallback) 1`] = `
Object {
  "_locale": "de",
  "_width": 1200,
  "html": "
<div data-dmid=search-container>
  <div data-dmid=search-content-container>
    <div data-dmid=search-body>
      <div
        data-dmid=search-fallback-container
        className=searchFallbackContainer__1dsx585
      >
        <h1 data-dmid=search-fallback-headline className=headline__12u4gl8>
          Ihre Suche nach „2ß9358ß239648” ergab leider keine Treffer
        </h1>
        <div
          data-dmid=search-fallback-info-container
          className=infoContainer__12ybgy7
        >
          <div
            data-dmid=search-fallback-info-container-content
            className=infoContainerContent__18pnl22
          >
            <div>content component 535460</div>
          </div>
          <div
            data-dmid=search-fallback-info-container-content
            className=infoContainerContent__18pnl22
          >
            <div>content component 535470</div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>;
",
  "styles": "
.headline__12u4gl8 {
  -webkit-hyphens: auto;
  -ms-hyphens: auto;
  hyphens: auto;
  font-family: dmbrand-medium, Arial, Helvetica, sans-serif;
  font-size: 1.75rem;
  color: #002878;
  text-align: center;
  margin: 1.5625rem 0 2.5rem;
}
.infoContainer__12ybgy7 {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-wrap: wrap;
  -webkit-box-lines: multiple;
  justify-content: space-around;
  -webkit-box-pack: justify;
  position: relative;
  margin-top: 0;
}
.infoContainerContent__18pnl22:first-of-type {
  padding-right: 2rem;
}
.infoContainerContent__18pnl22 {
  flex: 0 0 auto;
  width: 30%;
}
@media (max-width: 959px) {
  .infoContainerContent__18pnl22:first-of-type {
    padding-right: 0;
  }
  .infoContainerContent__18pnl22 {
    flex: 1 0 100%;
  }
  .searchFallbackContainer__1dsx585 {
    padding: 0 0.9375rem;
  }
}
",
}
`;

exports[`SearchContainer renders and matches snapshot when results are avaibable 1`] = `
Object {
  "_locale": "de",
  "_width": 1200,
  "html": "
<div data-dmid=search-container>
  <div data-dmid=search-nav className=searchNav__uw8rsa>
    <div>Filter-Nav</div>
  </div>
  <div data-dmid=search-content-container>
    <div data-dmid=search-header className=searchHeader__sjh0el>
      <h1
        data-dmid=search-header-headline
        className=searchHeaderHeadline__15pbijf
      >
        Suche nach „bio”
      </h1>
      <div
        data-dmid=search-tab-container
        className=searchTabContainer__umvgrv
      >
        <button
          type=submit
          data-dmattributes=active
          data-dmid=search-tab-PRODUCT
          className=searchTab__hr83d3
        >
          <div>PRODUCT Search Tab 42</div>
        </button>
        <button
          type=submit
          data-dmattributes
          data-dmid=search-tab-CONTENT
          className=searchTab__g9xlgf
        >
          <div>CONTENT Search Tab 42</div>
        </button>
      </div>
    </div>
    <div data-dmid=search-body>
      <div data-dmid=search-results>
        <div>
          <h1>Suche nach bio</h1>
          <div>
            <span>result 1</span>
            <span>result 2</span>
            <span>result 3</span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>;
",
  "styles": "
.searchNav__uw8rsa {
  width: 13.75em;
  padding: 0 3.125rem 0 0;
  display: block;
}
.searchHeader__sjh0el {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-flow: column nowrap;
}
.searchTabContainer__umvgrv {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  align-items: center;
  -webkit-box-align: center;
  justify-content: center;
  -webkit-box-pack: center;
  margin-bottom: 2.5rem;
}
.searchHeaderHeadline__15pbijf {
  font-family: dmbrand-medium, Arial, Helvetica, sans-serif;
  font-size: 2.625rem;
  color: #002878;
  text-align: center;
  margin: 0 0 1.25rem !important;
}
.searchTab__hr83d3:focus {
  outline: none;
}
.searchTab__hr83d3:after {
  box-sizing: border-box;
  display: block;
  content: \\" \\";
  position: absolute;
  width: 1em;
  height: 1em;
  border: 2px solid #002878;
  z-index: 1;
  background-color: #002878;
  top: 2em;
  margin-left: -6px;
  left: 50%;
  transform: rotate(45deg);
}
.searchTab__hr83d3 {
  background-color: #002878;
  height: 2.5rem;
  line-height: 2.7em;
  color: #ffffff;
  border: 0;
  cursor: pointer;
  position: relative;
  z-index: 1;
  margin: 0.3125rem;
  border-radius: 1.5625rem;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  align-items: center;
  -webkit-box-align: center;
  justify-content: center;
  -webkit-box-pack: center;
  font-size: 1rem;
  font-family: dmbrand-medium, Arial, Helvetica, sans-serif;
  padding: 0 0.9375rem;
}
.searchTab__g9xlgf:focus {
  outline: none;
}
.searchTab__g9xlgf:after {
  box-sizing: border-box;
  display: none;
  content: \\" \\";
  position: absolute;
  width: 1em;
  height: 1em;
  border: 2px solid #002878;
  z-index: 1;
  background-color: #002878;
  top: 2em;
  margin-left: -6px;
  left: 50%;
  transform: rotate(45deg);
}
.searchTab__g9xlgf {
  background-color: #e5e9f1;
  height: 2.5rem;
  line-height: 2.7em;
  color: #002878;
  border: 0;
  cursor: pointer;
  position: relative;
  z-index: 1;
  margin: 0.3125rem;
  border-radius: 1.5625rem;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  align-items: center;
  -webkit-box-align: center;
  justify-content: center;
  -webkit-box-pack: center;
  font-size: 1rem;
  font-family: dmbrand-medium, Arial, Helvetica, sans-serif;
  padding: 0 0.9375rem;
}
@media (max-width: 959px) {
  .searchNav__uw8rsa {
    display: none;
    width: 0;
    padding: 0;
  }
  .searchHeader__sjh0el {
    padding: 0 0.9375rem;
  }
}
",
}
`;
